### DISCUSSION TITLE
Hue deformation i LAB colorspace

### DISCUSSION THREAD

**Discussion by SteinarVatne** [ImageMagick/ImageMagick#5595] [Help]
I'm not a regular ImageMagick user, but I'm looking  for a quick and robust way to reproduce a Photoshop workflow that I developed for color calibration of images. I have an image that contains a simple calibration palette of 4 colors (red, green, blue and grey). These colors have also been measured separately, and the result is given in LAB values. My first attempt in Photoshop was to use a "curves" adjustment layer where I map input values (averaged eyedropper points on the calibration surfaces) to output values (the values from physical measurement). However, I soon realised that this wasn't going to work out, because the different calibration colors asked for channel adjustments that were to some degree mutually exclusive.

I then had the idea that if I could calculate the hue/angle in the colorwheel for each measurement and surface, then I could "twist" the surfaces to the correct hue first, and then try the curves adjustment again. This seemed to do the trick, as I was now able to match all the values within the required tolerance. This "twisting" is the main part that I was hoping to recreate in ImageMagick. In Photoshop, this was done with a Hue/Saturation adjustment layer, where you can modify the center and falloff range of a number of "channels", and then shift these left or right around the colorwheel.

Is there some manipulation available in ImageMagick that can do this? The -modulate seems to only shift the entire range? I guess the Hald 3D CLUT could do it, but then I need to produce the CLUT image elsewhere?

**Comment by snibgo** [comment]
For most definitions of "hue", it is easy to add (or subtract) a hue-shift to all pixels, so all colours will shift by the same amount.

If you want a different hue-shift for different hues, the task isn't as simple. But if you have a Photoshop process that does what you want, you can use a Hald CLUT to easily repeat the process.

1. In IM, create an "identity" (aka "null") Hald Clut:
```
magick hald:12 h.png
```
(12 is usually fine. Use larger numbers eg 16 for larger but more precise CLUTs, or smaller eg 8 for the opposite.)

2. In Photoshop (or whatever program you want) read h.png and apply the process that does what you want. Save the result as h2.png.

3. Now you can apply the Hald CLUT to any image you want, like this:
```
magick in.ext h2.png -hald-clut out.ext
```
Use any filenames you want.

Does that help?

**Comment by SteinarVatne** [comment]
Yes, I realised after thinking some more that the 3D CLUT thing will almost certainly work. I was in part hoping to automate the discovery of the actual parameters to the transformation, which will not be possible by this route though.

**Comment by snibgo** [comment]
If you have two images, where one has been transformed to another by changing pixel colours, there are many methods for estimating the transformation, including: matching histograms, matching gain and bias (or just gain, or just bias), best-fit bezier or polynomial per channel, best-fit 3x3 matrix, sparse cluts (one per channel), sparse Hald clut.

These methods make different assumptions about the nature of the transformation. Of the methods, sparse Hald clut is the most flexible as it should transform exactly all the colours present in the transformed image. However, I haven't found an IM solution that finds the transformation in a reasonable time.

See, for example, [Colour checker charts](http://im.snibgo.com/colchkcha.htm).

**Comment by SteinarVatne** [comment]
I'm probably not explaining it very well; I meant that I would like to discover the parameters for the Hue/Saturation transform (what points that need to be rotated, with which falloff range, and by how much) by analysing the difference between actual pixel colors and measured colors. This will perhaps require more of a programming/script-interface... I did this partly in Maxscript (3D Studio MAX) for my proof-of-concept, but I was hoping to put together something a bit more streamlined, without this jumping back and forth between 3DS MAX, Photoshop and Windows Calculator.

**Comment by SteinarVatne** [comment]
Actually, something like what you describe could perhaps find such a transform directly, but given that I only have a very small number of before/after samples, I'm guessing they won't work very well in this instance.

**Comment by snibgo** [comment]
That depends on the nature of the transformation. For example, if it affects only the chroma and hue channels of LCHab colorspace, and the transformations of those channels are independant, then the problem is fairly simple. And if the transformations are linear, ie `out = in*a + b`, then not many samples are needed to find `a` and `b`.

**Reply by SteinarVatne** [reply]
I'm out of my depths here, but the respective LAB-channels are not chroma and hue, just A and B, where A goes from -127 to 127 (green <-> magenta?) and B also from -127 to 127 (yellow <-> blue?) The L(ightness) channel does not have to match. But like I said, it's not possible to get it right by just manipulating the levels of those channels independantly. Hence the need for a hue cycling/deformation, which in my understanding represents a transformation where the change in one channel depends on the current value of the other channel?

However, it's not clear to me exactly what is causing this transformation of colors that I seek to undo, there are too many steps and variables that I have no knowledge of or control over. As such, I don't know for sure what is the "correct" transform, but the way I'm doing  it now, the hue-twisting is uniform (the same angle in degrees) over all chroma-levels, where chroma-strength is defined by distance from the center of the colorwheel. You could perhaps accomplish a similar result by treating the samples as points instead of radial lines, and maybe even eliminate the need for a final curves adjustment by shifting them in chroma strength simultaneously?

**Comment by fmw42** [comment]
In Imagemagick, you would shift the whole image channel and use a gradient ramp mask to limit the range.

What platform/os are you using?  I have some bash unix script that might help. See my huemap, modulatecolor1, and modulatecolor2 scripts at http://www.fmwconcepts.com/imagemagick

**Reply by SteinarVatne** [reply]
Thx, I'm on Windows. Your modulatecolor2 seems to be almost what I'd need for twisting the hues, except that in Photoshop, I can set the tolerance/ramp thing individually for either side of the center. Even so, I could possibly get away with using the lesser range for both. I guess I was hoping for a solution that would be simpler overall than this will be, I'm probably better off just setting up some Photoshop actions and updating them manually when I need a new calibration. But thank you for the answers!

**Comment by snibgo** [comment]
> ... but the respective LAB-channels are not chroma and hue,...

I referred to the LCHab colorspace which has channels Lightness, Chroma and Hue. It is L\*a\*b\* expressed as a cylinder.

**Reply by SteinarVatne** [reply]
Like I said, I'm out of my depth! :-) It is however the L*a*b space I need to work with, so perhaps that is another hurdle?

**Comment by SteinarVatne** [comment]
I think I'll stick with Photoshop for working out the transform, but there is probably a good chance that a Hald 3D CLUT operation in ImageMagick will outperform a Photoshop batch process for hundreds of very large images?

**Comment by snibgo** [comment]
I don't use Photoshop, and have no idea how fast that would be.

For many images with IM, write the Hald CLUT to a .miff or .mpc file, as IM can read these more quickly than a .PNG file.
